std/file.d

```diff
@@ -868,20 +849,202 @@ version(Windows) private void writeImpl(const(char)[] name, const(FSChar)* namez
+
+        if (numWritten != cnt)
             break;
-        sum += numwritten;
+        sum += numWritten;
     }
-    cenforce(sum == size && CloseHandle(h), name, namez);
+
+    version(Posix)
+        cenforce(sum == size && core.sys.posix.unistd.close(fd) == 0, name, namez);
+    else version(Windows)
+        cenforce(sum == size && CloseHandle(fd), name, namez);
+    else
+        static assert(0, "Unsupported OS");
 }
 
+
+/++
+    Creates a file with a random name in $(LREF tempDir) and returns the name
+    of the file. If data is passed in, then the file is written with that data;
+    otherwise, it's empty.
+
+    The idea is that this creates a temporary file for testing or whatever
+    other purpose a temporary file might be needed for. However, it will only
+    be deleted if it's explicitly deleted or if the OS decides to delete it as
+    some OSes do with files in temp directories on startup or shutdown. So,
+    unlike $(REF File.tmpfile, std, stdio), the file has a name, and it can have
+    all of the things done to it that one might typically do with a file
+    (including reopening it after closing it).
+
+    The file is created with R/W permissions. On POSIX systems, the permissions
+    are restricted to the current user, though the effective permissions are
+    modified by the process' umask in the usual way.
+
+    $(REF rndGen, std, random) is used as the random number generator.
+
+    Params:
+        prefix = Prefix to the random portion of the file name.
+        suffix = Suffix to the random portion of the file name (e.g. for a file
+                 extension).
+        buffer = Data to populate the file with.
+
+    Returns:
+        The name of the generated file.
+
+    Throws:
+        $(LREF FileException) if it fails to create the file.
+
+    See_Also:
+        $(REF File.tmpfile, std, stdio)
+  +/
+string createTempFile(const void[] buffer = null) @safe
+{
+    return createTempFile(null, null, buffer);
+}
+
+/// Ditto
+string createTempFile(const(char)[] prefix, const(char)[] suffix, const void[] buffer = null) @trusted
+{
+    import std.path : absolutePath, baseName, buildPath, dirName;
+
+    static string genTempName(const(char)[] prefix, const(char)[] suffix)
+    {
+        import std.ascii : digits, letters;
+        import std.random : choice, rndGen;
+        import std.range : chain;
+        import std.string : representation;
+
+        auto name = new char[](prefix.length + 15 + suffix.length);
+        name[0 .. prefix.length] = prefix;
+        name[$ - suffix.length .. $] = suffix;
+
+        auto random = &rndGen();
+        rndGen.popFront();
+
+        auto chars = chain(letters.representation, digits.representation);
+        foreach (ref c; name[prefix.length .. $ - suffix.length])
+        {
+            c = choice(chars);
+            random.popFront();
+        }
+
+        return buildPath(tempDir, name);
+    }
+
+    while (1)
```

Comment:
Using `1` as `true` is a C-ism (which does not have a `bool` type). (Since D has a `bool` type, numbers should be used to indicate quantity, and 1 here does not indicate the quantity of anything.) Please use `while (true)` in D.